### PR TITLE
Add support for multiple losses with tf.keras

### LINKED ISSUE #386: Support for multiple losses with tf.keras
If I understand correctly, tf-federated only supports single loss function. Is there any timeline on when the support for multiple losses will be added?

```
File "/home/.local/lib/python3.5/site-packages/tensorflow_federated/python/learning/model_utils.py", line 456, in __init__
    raise NotImplementedError('only a single loss functions is supported')
```

### PR DESCRIPTION
**Pull request by Squadrick**
Fixes #386

### CODE REVIEW

**Comment by Squadrick** [comment]
Still need to add more test cases, and update docs.

**Comment by Squadrick** [comment]
@ZacharyGarrett Do you guys use any linting or formatting tools like `pylint` or `yapf`? I didn't find anything in `CONTRIBUTING.md`.

For testing: 
```
bazel test //tensorflow_federated/python/learning:keras_utils_test
```

**Comment by Squadrick** [comment]
Also, I can add support for `loss_weights`. Do you think it's better if I add that to this PR, or make a separate PR once this is merged?

**Comment by ZacharyGarrett** [comment]
@Squadrick we do use both `pylint` and `yapf`, we'll add notes to the `CONTRIBUTING.md` (thanks for calling that out!).

This is looking really good. If you'd like to go ahead and add support for `loss_weights`, that would be great!

**Comment by Squadrick** [comment]
@ZacharyGarrett Added support for `loss_weights`. Could you take a look?

**Comment by Squadrick** [comment]
Also, should I squash all my commits into a single one?

File: tensorflow_federated/python/learning/keras_utils.py [hunk 1]
```diff
@@ -230,7 +277,7 @@ def report(accumulators):
 class _KerasModel(model_lib.Model):
   """Internal wrapper class for tf.keras.Model objects."""
 
-  def __init__(self, inner_model, dummy_batch, loss_fn, metrics):
+  def __init__(self, inner_model, dummy_batch, loss_fns, loss_weights, metrics):
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]:
```suggestion
  def __init__(self, inner_model, dummy_batch, loss_fns, loss_weights=None, metrics=None):
```

Lets add defaults for everything after `loss_fns`

File: tensorflow_federated/python/learning/keras_utils.py [hunk 2]
```diff
@@ -246,7 +293,24 @@ def _tensor_spec_with_undefined_batch_dim(tensor):
         _tensor_spec_with_undefined_batch_dim, dummy_batch)
 
     self._keras_model = inner_model
-    self._loss_fn = loss_fn
+    self._loss_fns = loss_fns
+
+    if isinstance(loss_weights, collections.Mapping):
+      print(inner_model.output_names)
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]:
```suggestion
```

File: tensorflow_federated/python/learning/keras_utils.py [hunk 3]
```diff
@@ -246,7 +293,24 @@ def _tensor_spec_with_undefined_batch_dim(tensor):
...
+      print(loss_weights.keys())
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]:
```suggestion
```

File: tensorflow_federated/python/learning/keras_utils.py [hunk 4]
```diff
@@ -98,7 +100,17 @@ def from_keras_model(keras_model,
       The values of the tensors are not important and can be filled with any
       reasonable input value.
     loss: A callable that takes two batched tensor parameters, `y_true` and
-      `y_pred`, and returns the loss.
+      `y_pred`, and returns the loss. If the model has multiple outputs, you
+      can use a different loss on each output by passing a dictionary or a list
+      of losses. The loss value that will be minimized by the model will then
+      be the sum of all individual losses.
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]:
```suggestion
      be the sum of all individual losses, each weighted by `loss_weights`.
```

File: tensorflow_federated/python/learning/keras_utils.py [hunk 5]
```diff
@@ -111,14 +123,49 @@ def from_keras_model(keras_model,
...
   py_typecheck.check_type(keras_model, tf.keras.Model)
+  py_typecheck.check_type(loss, (tf.keras.losses.Loss,
+                                 collections.Sequence,
+                                 collections.Mapping))
+
+  if loss_weights is not None:
+    py_typecheck.check_type(loss, (collections.Sequence,
+                                   collections.Mapping))
+
+  if isinstance(loss, collections.Sequence):
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]:
```suggestion
  if isinstance(loss, (collections.Mapping, collections.Sequence)):
```

File: tensorflow_federated/python/learning/keras_utils.py [hunk 6]
```diff
@@ -111,14 +123,49 @@ def from_keras_model(keras_model,
...
+    if len(loss) != len(keras_model.outputs):
+      raise ValueError('`keras_model` must have equal number of '
+                       'outputs and losses')
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]: TFF generally aims to include the parameters in the error string for easier debugging, this is probably different than traditional Python.

In this case, perhaps we could add the string `'loss: {}\noutputs: {}'.format(loss, keras_model.outputs)`. Same fro the error raised below.

File: tensorflow_federated/python/learning/keras_utils.py [hunk 7]
```diff
@@ -111,14 +123,49 @@ def from_keras_model(keras_model,
...
     raise ValueError('`keras_model` must not be compiled. Use '
                      'from_compiled_keras_model() instead.')
+
   dummy_tensors = _preprocess_dummy_batch(dummy_batch)
   if optimizer is None:
+    if isinstance(loss, collections.Mapping):
+      loss_functions = []
+      for name in keras_model.output_names:
+        if name not in loss:
+          logging.warning(
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]: TFF tries to be strict and avoid assumptions on user input. Lets change this to raise an error instead, and include the names of the outputs and losses so the user can easily determine what was missing.

File: tensorflow_federated/python/learning/keras_utils.py [hunk 8]
```diff
@@ -111,14 +123,49 @@ def from_keras_model(keras_model,
...
+              'Output {0} missing from loss dictionary. We assume '
+              'this was done on purpose. The fit and evaluate APIs will not be '
+              'expecting any data to be passed to {0}.'.format(name))
+        loss_functions.append(loss[name])
+
+    elif isinstance(loss, collections.Sequence):
+      loss_functions = loss
+
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils.py]: Is `yapf` adding the newlines before each new condition branch? This isn't common in TFF, if possible lets remove the empty newlines.

File: tensorflow_federated/python/learning/keras_utils_test.py [hunk 9]
```diff
@@ -451,6 +451,108 @@ def test_keras_model_and_optimizer(self):
...
+    with self.subTest("loss_weights_assert_fail"):
+      with self.assertRaises(ValueError):
+        _ = keras_utils.from_keras_model(
+          keras_model=keras_model,
+          dummy_batch=dummy_batch,
+          loss=[tf.keras.losses.MeanSquaredError(),
+                tf.keras.losses.MeanSquaredError(),
+                tf.keras.losses.MeanSquaredError()],
+          loss_weights=[0.1, 0.2])
+
```
Comment (ZacharyGarrett) [inline tensorflow_federated/python/learning/keras_utils_test.py]: Lets add one more test with a dict of loss_weights that doesn't have the same keys are the outputs.

**Comment by ZacharyGarrett** [comment]
> Also, should I squash all my commits into a single one?

Either is fine. In the end, it will get squashed before merging.

TFF uses an internal tool to automate merging PRs after they have been approved. It does various things such as ensuring all unittests pas, no pylint warnings, etc. If I understand correctly, it will squash.

**Comment by Squadrick** [comment]
Made the suggested changes. Also, ran `yapf` with `--style='{based_on_style: chromium, indent_width:2}'`. I guessed that this was the style used by TFF. Let me know if I used wrong formatting style.

**Comment by Squadrick** [comment]
@ZacharyGarrett Thanks for reviewing the PR!
